##

## installing necessary libraries and modules

In [ ]:

! pip install pandas
! python -m pip install "pymongo[srv]"
! pip install mysql-connector-python

Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\91797\AppData\Local\Programs\Python\Python312\Scripts\pip.exe\__main__.py", line 4, in <module>
ModuleNotFoundError: No module named 'pip'
c:\Users\91797\AppData\Local\Programs\Python\Python312\python.exe: No module named pip
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "C:\Users\91797\AppData\Local\Programs\Python\Python312\Scripts\pip.exe\__main__.py", line 4, in <module>
ModuleNotFoundError: No module named 'pip'


## importing the requirements

In [ ]:


import pandas as pd
import numpy as np
from pymongo.mongo_client import MongoClient
from pymongo.server_api import ServerApi
import mysql.connector


import matplotlib.pyplot as plt
import seaborn as sns
import warnings


warnings.filterwarnings("ignore")

%matplotlib inline

In [5]:
# reading the CSV data file

df = pd.read_csv("C:\\Users\\91797\\OneDrive\\Desktop\\app\\census_2011.csv", header = 0)

In [6]:
df

District code                   State name              District name  \
0                1            JAMMU AND KASHMIR                    Kupwara   
1                2            JAMMU AND KASHMIR                     Badgam   
2                3            JAMMU AND KASHMIR                Leh(Ladakh)   
3                4            JAMMU AND KASHMIR                     Kargil   
4                5            JAMMU AND KASHMIR                      Punch   
..             ...                          ...                        ...   
635            636                  PONDICHERRY                       Mahe   
636            637                  PONDICHERRY                   Karaikal   
637            638  ANDAMAN AND NICOBAR ISLANDS                   Nicobars   
638            639  ANDAMAN AND NICOBAR ISLANDS  North  AND Middle Andaman   
639            640  ANDAMAN AND NICOBAR ISLANDS              South Andaman   

     Population      Male    Female  Literate  Male_Literate  Female_Literate  \
0      870354.0  474190.0  396164.0  439654.0       282823.0         156831.0   
1      753745.0       NaN  355704.0  335649.0       207741.0         127908.0   
2      133487.0   78971.0   54516.0   93770.0        62834.0          30936.0   
3      140802.0       NaN   63017.0       NaN        56301.0          29935.0   
4           NaN  251899.0  224936.0  261724.0       163333.0          98391.0   
..          ...       ...       ...       ...            ...              ...   
635     41816.0   19143.0   22673.0   36470.0        16610.0          19860.0   
636    200222.0   97809.0  102413.0  154916.0        79903.0          75013.0   
637     36842.0   20727.0       NaN   25332.0        15397.0           9935.0   
638    105597.0   54861.0   50736.0   78683.0        43186.0          35497.0   
639         NaN  127283.0  110859.0  190266.0       105794.0          84472.0   

          SC  ...  Power_Parity_Rs_90000_150000  Power_Parity_Rs_45000_150000  \
0     1048.0  ...                          94.0                         588.0   
1        NaN  ...                         126.0                         562.0   
2      488.0  ...                          46.0                         122.0   
3       18.0  ...                          27.0                         114.0   
4      556.0  ...                          78.0                         346.0   
..       ...  ...                           ...                           ...   
635    144.0  ...                        2316.0                        4309.0   
636  35348.0  ...                        1063.0                        2408.0   
637      0.0  ...                         685.0                        1895.0   
638      0.0  ...                         685.0                        1895.0   
639      0.0  ...                        1371.0                        3020.0   

     Power_Parity_Rs_150000_240000  Power_Parity_Rs_240000_330000  \
0                             71.0                          101.0   
1                             72.0                           89.0   
2                             15.0                           22.0   
3                             12.0                           18.0   
4                             35.0                           50.0   
..                             ...                            ...   
635                         1370.0                          838.0   
636                          665.0                          340.0   
637                          212.0                          134.0   
638                          212.0                            NaN   
639                          649.0                          368.0   

     Power_Parity_Rs_150000_330000  Power_Parity_Rs_330000_425000  \
0                            172.0                           74.0   
1                            161.0                           96.0   
2                              NaN                           20.0   
3                

### Shape of the data

In [7]:
df.shape

(640, 118)

The dataset contains 640 rows and 118 columns.

 ### Summary of the dataset :

In [8]:
# Displaying summary for a dataframe

df.describe()

District code    Population          Male        Female      Literate  \
count     640.000000  6.100000e+02  6.100000e+02  6.070000e+02  6.040000e+02   
mean      320.500000  1.852620e+06  9.685726e+05  9.139882e+05  1.190550e+06   
std       184.896367  1.435578e+06  7.926455e+05  7.479293e+05  1.050630e+06   
min         1.000000  8.004000e+03  4.414000e+03  3.590000e+03  4.436000e+03   
25%       160.750000  8.148625e+05  4.180382e+05  3.986990e+05  4.796545e+05   
50%       320.500000  1.557367e+06  7.939620e+05  7.556020e+05  9.793575e+05   
75%       480.250000  2.565248e+06  1.318615e+06  1.253832e+06  1.602260e+06   
max       640.000000  1.000978e+07  5.865078e+06  5.195070e+06  8.227161e+06   

       Male_Literate  Female_Literate            SC       Male_SC  \
count   6.090000e+02     6.130000e+02  6.050000e+02  6.170000e+02   
mean    6.739088e+05     5.129481e+05  3.177854e+05  1.624509e+05   
std     5.683384e+05     4.801441e+05  3.172010e+05  1.625607e+05   
min     2.614000e+03     1.822000e+03  0.000000e+00  0.000000e+00   
25%     2.695820e+05     2.009450e+05  8.799100e+04  4.251300e+04   
50%     5.507160e+05     4.048610e+05  2.463370e+05  1.252140e+05   
75%     9.203140e+05     6.568420e+05  4.550620e+05  2.284310e+05   
max     4.591396e+06     3.635765e+06  2.464032e+06  1.266504e+06   

          Female_SC  ...  Power_Parity_Rs_90000_150000  \
count  6.080000e+02  ...                    607.000000   
mean   1.544700e+05  ...                    792.028007   
std    1.537418e+05  ...                   1059.010710   
min    0.000000e+00  ...                      0.000000   
25%    4.411825e+04  ...                    236.000000   
50%    1.181440e+05  ...                    507.000000   
75%    2.163095e+05  ...                    939.000000   
max    1.197528e+06  ...                  10334.000000   

       Power_Parity_Rs_45000_150000  Power_Parity_Rs_150000_240000  \
count                    610.000000                     611.000000   
mean                    1676.009836                     288.960720   
std                     1669.641260                     641.765892   
min                        0.000000                       0.000000   
25%                      588.000000                      59.000000   
50%                     1196.000000                     144.000000   
75%                     2224.250000                     293.000000   
max                    13483.000000                   10835.000000   

       Power_Parity_Rs_240000_330000  Power_Parity_Rs_150000_330000  \
count                     605.000000                     616.000000   
mean                      213.919008                     513.099026   
std                       370.466703                     985.702797   
min                         0.000000                       0.000000   
25%                        23.000000                      93.500000   
50%                       107.000000                     274.000000   
75%                       253.000000                     562.250000   
max                      3595.000000                   14430.000000   

       Power_Parity_Rs_330000_425000  Power_Parity_Rs_425000_545000  \
count                     607.000000                     610.000000   
mean                      197.375618                     264.191803   
std                       433.595214                     599.182627   
min                         0.000000                       0.000000   
25%                        19.000000                      21.000000   
50%                        84.000000                      84.000000   
75%                       216.000000                     293.000000   
max                      5027.000000                    7597.000000   

       Power_Parity_Rs_330000_545000  Power_Parity_Above_Rs_545000  \
count                     617.000000                    610.000000   
mean                      463.019449                    283.934426   
std                 

### Check Datatypes in the dataset

In [9]:
# Check Null and Dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 640 entries, 0 to 639
Columns: 118 entries, District code to Total_Power_Parity
dtypes: float64(115), int64(1), object(2)
memory usage: 590.1+ KB


### Check for all columns

In [10]:
df.columns

Index(['District code', 'State name', 'District name', 'Population', 'Male',
       'Female', 'Literate', 'Male_Literate', 'Female_Literate', 'SC',
       ...
       'Power_Parity_Rs_90000_150000', 'Power_Parity_Rs_45000_150000',
       'Power_Parity_Rs_150000_240000', 'Power_Parity_Rs_240000_330000',
       'Power_Parity_Rs_150000_330000', 'Power_Parity_Rs_330000_425000',
       'Power_Parity_Rs_425000_545000', 'Power_Parity_Rs_330000_545000',
       'Power_Parity_Above_Rs_545000', 'Total_Power_Parity'],
      dtype='object', length=118)

### Task -1 : Renaming the column Names

In [11]:
# Renaming the columns

column_headers = [
    {'State name':'State/UT'},
    {'District name':'District'},
    {'Male_Literate':'Literate_Male'},
    {'Female_Literate':'Literate_Female'},
    {'Rural_Households':'Households_Rural'},
    {'Urban_Households':'Households_Urban'},
    {'Age_Group_0_29':'Young_and_Adult'},
    {'Age_Group_30_49':'Middle_Aged'},
    {'Age_Group_50':'Senior_Citizen'},
    {'Age not stated':'Age_Not_Stated'}
    ]

def rename_col(headers,df):
    for i in headers:
        df.rename(columns = i, inplace = True)

rename_col(column_headers, df)

In [12]:
df.head(5)

District code           State/UT     District  Population      Male  \
0              1  JAMMU AND KASHMIR      Kupwara    870354.0  474190.0   
1              2  JAMMU AND KASHMIR       Badgam    753745.0       NaN   
2              3  JAMMU AND KASHMIR  Leh(Ladakh)    133487.0   78971.0   
3              4  JAMMU AND KASHMIR       Kargil    140802.0       NaN   
4              5  JAMMU AND KASHMIR        Punch         NaN  251899.0   

     Female  Literate  Literate_Male  Literate_Female      SC  ...  \
0  396164.0  439654.0       282823.0         156831.0  1048.0  ...   
1  355704.0  335649.0       207741.0         127908.0     NaN  ...   
2   54516.0   93770.0        62834.0          30936.0   488.0  ...   
3   63017.0       NaN        56301.0          29935.0    18.0  ...   
4  224936.0  261724.0       163333.0          98391.0   556.0  ...   

   Power_Parity_Rs_90000_150000  Power_Parity_Rs_45000_150000  \
0                          94.0                         588.0   
1                         126.0                         562.0   
2                          46.0                         122.0   
3                          27.0                         114.0   
4                          78.0                         346.0   

   Power_Parity_Rs_150000_240000  Power_Parity_Rs_240000_330000  \
0                           71.0                          101.0   
1                           72.0                           89.0   
2                           15.0                           22.0   
3                           12.0                           18.0   
4                           35.0                           50.0   

   Power_Parity_Rs_150000_330000  Power_Parity_Rs_330000_425000  \
0                          172.0                           74.0   
1                          161.0                           96.0   
2                            NaN                           20.0   
3                           30.0                           19.0   
4                           85.0                           59.0   

   Power_Parity_Rs_425000_545000  Power_Parity_Rs_330000_545000  \
0                           10.0                           84.0   
1                           28.0                          124.0   
2                            NaN                            NaN   
3                            3.0                           22.0   
4                            8.0                           67.0   

   Power_Parity_Above_Rs_545000  Total_Power_Parity  
0                          15.0              1119.0  
1                          18.0              1066.0  
2                          17.0               242.0  
3                           7.0               214.0  
4                          12.0               629.0  

[5 rows x 118 columns]

### Task-2 : Rename State/UT Names

- The State/UT names are in all caps in the census data, For uniformity across datasets we use the names so that only the first character of each word in the name is in upper case and the rest are in lower case.

- However, if the word is “and” then it should be all lowercase.if you found & symbol replace it with ‘and’


In [13]:
# We have two for loops. one is outer loop & another is Inner loop
# Where Outer for loop to get the state names and  Inner for loop to check if the word is "and" else capitalize the word

def rename_state_ut(df):
    for i in range(len(df)):
        x = df.loc[i,'State/UT'].lower().split()
        for j in range(len(x)):
            if x[j] != 'and':
                x[j] = x[j].capitalize()
        df.loc[i,'State/UT'] = " ".join(x)

rename_state_ut(df)

### Define numerical & Categorical columns

In [14]:
numeric_features =[feature for feature in df.columns if df[feature].dtype != 'O']
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'O']

#Print columns
print('We have {} numerical features : {}'.format(len(numeric_features),numeric_features))
print('We have {} categorical features : {}'.format(len(categorical_features),categorical_features))

We have 116 numerical features : ['District code', 'Population', 'Male', 'Female', 'Literate', 'Literate_Male', 'Literate_Female', 'SC', 'Male_SC', 'Female_SC', 'ST', 'Male_ST', 'Female_ST', 'Workers', 'Male_Workers', 'Female_Workers', 'Main_Workers', 'Marginal_Workers', 'Non_Workers', 'Cultivator_Workers', 'Agricultural_Workers', 'Household_Workers', 'Other_Workers', 'Hindus', 'Muslims', 'Christians', 'Sikhs', 'Buddhists', 'Jains', 'Others_Religions', 'Religion_Not_Stated', 'LPG_or_PNG_Households', 'Housholds_with_Electric_Lighting', 'Households_with_Internet', 'Households_with_Computer', 'Households_Rural', 'Households_Urban', 'Households', 'Below_Primary_Education', 'Primary_Education', 'Middle_Education', 'Secondary_Education', 'Higher_Education', 'Graduate_Education', 'Other_Education', 'Literate_Education', 'Illiterate_Education', 'Total_Education', 'Young_and_Adult', 'Middle_Aged', 'Senior_Citizen', 'Age_Not_Stated', 'Households_with_Bicycle', 'Households_with_Car_Jeep_Van', 'Ho

In [15]:
df[numeric_features].head(5)

District code  Population      Male    Female  Literate  Literate_Male  \
0              1    870354.0  474190.0  396164.0  439654.0       282823.0   
1              2    753745.0       NaN  355704.0  335649.0       207741.0   
2              3    133487.0   78971.0   54516.0   93770.0        62834.0   
3              4    140802.0       NaN   63017.0       NaN        56301.0   
4              5         NaN  251899.0  224936.0  261724.0       163333.0   

   Literate_Female      SC  Male_SC  Female_SC  ...  \
0         156831.0  1048.0   1046.0        2.0  ...   
1         127908.0     NaN    343.0       25.0  ...   
2          30936.0   488.0    444.0       44.0  ...   
3          29935.0    18.0     12.0        6.0  ...   
4          98391.0   556.0    406.0        NaN  ...   

   Power_Parity_Rs_90000_150000  Power_Parity_Rs_45000_150000  \
0                          94.0                         588.0   
1                         126.0                         562.0   
2                          46.0                         122.0   
3                          27.0                         114.0   
4                          78.0                         346.0   

   Power_Parity_Rs_150000_240000  Power_Parity_Rs_240000_330000  \
0                           71.0                          101.0   
1                           72.0                           89.0   
2                           15.0                           22.0   
3                           12.0                           18.0   
4                           35.0                           50.0   

   Power_Parity_Rs_150000_330000  Power_Parity_Rs_330000_425000  \
0                          172.0                           74.0   
1                          161.0                           96.0   
2                            NaN                           20.0   
3                           30.0                           19.0   
4                           85.0                           59.0   

   Power_Parity_Rs_425000_545000  Power_Parity_Rs_330000_545000  \
0                           10.0                           84.0   
1                           28.0                          124.0   
2                            NaN                            NaN   
3                            3.0                           22.0   
4                            8.0                           67.0   

   Power_Parity_Above_Rs_545000  Total_Power_Parity  
0                          15.0              1119.0  
1                          18.0              1066.0  
2                          17.0               242.0  
3                           7.0               214.0  
4                          12.0               629.0  

[5 rows x 116 columns]

In [16]:
df[categorical_features].head(5)

State/UT     District
0  Jammu and Kashmir      Kupwara
1  Jammu and Kashmir       Badgam
2  Jammu and Kashmir  Leh(Ladakh)
3  Jammu and Kashmir       Kargil
4  Jammu and Kashmir        Punch

### Task -3 New State/UT formation with encoding handling

- In 2014 Telangana was formed after it split from Andhra Pradesh, The districts that were included in Telangana are stored in Data/Telangana.txt . Read the text file and Rename the State/UT From “Andhra Pradesh” to “Telangana” for the given districts.

- In 2019 Ladakh was formed after it split from Jammu and Kashmir, which included the districts Leh and Kargil. Rename the State/UT From “Jammu and Kashmir” to “Ladakh” for the given districts.

In [19]:
# Reading the txt file

with open('C:\\Users\\91797\\OneDrive\\Desktop\\app\\Telangana.txt', 'r') as telangana:
    telangana_districts = telangana.read()
telangana_districts = telangana_districts.split("\n")

### Setting the values as Telangana if the districts are in the telangana districts list.

In [ ]:
def telangana(df):
    for i in range(len(df)):
        if df.loc[i,'District'] in telangana_districts:
            df.loc[i,'State/UT'] = "Telangana"

telangana(df)

### Setting the values as Ladakh if the districts are in the ladakh list .

In [ ]:
def ladakh(df):
    ladakh = ['Leh', 'Kargil']
    for district in ladakh:
        for i in range(len(df)):
            if district in df.loc[i,'District']:
                df.loc[i,'State/UT'] = "Ladakh"

ladakh(df)

### Task 4: Find and process Missing Data

#### Finding and storing the Percentage of missing data before cleaning for each column in a separate dictionary.

In [ ]:
def missingDataPercent(df) :
    missing_data = {}
    columns = df.columns
    for i in columns:
        count = 0
        count += len(list(filter(lambda x : x == True, list(df[i].isnull()))))
        missing_data[i] = round((count * 100)/len(df), 2)

    return missing_data

missing_data_before_cleaning = missingDataPercent(df)

#### Function to return if the cell contains null value or not :

In [ ]:
def null_values_filter(df,i,column):
    return list(df.loc[[i],column].isnull())[0]

#### Filling up the Population column :

In [ ]:
def fill_population(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Population') == True:
            if null_values_filter(df,i,'Male') == null_values_filter(df,i,'Female') == False:
                df.loc[i,"Population"] = int(df.loc[i,"Male"] + df.loc[i, "Female"])
            elif null_values_filter(df,i,'Young_and_Adult') == null_values_filter(df,i,'Middle_Aged') == null_values_filter(df,i,"Senior_Citizen") == null_values_filter(df,i,'Age_Not_Stated') == False:
                df.loc[i,'Population'] = int(df.loc[i,"Young_and_Adult"] + df.loc[i,"Middle_Aged"] + df.loc[i, "Senior_Citizen"] + df.loc[i,"Age_Not_Stated"])
            elif null_values_filter(df,i,'Hindus') == null_values_filter(df,i,'Muslims') == null_values_filter(df,i,'Christians') == null_values_filter(df,i,'Sikhs') == null_values_filter(df,i,'Buddhists') == null_values_filter(df,i,'Jains') == null_values_filter(df,i,'Others_Religions') == null_values_filter(df,i,'Religion_Not_Stated') == False:
                df.loc[i,'Population'] = int(df.loc[i,'Hindus'] + df.loc[i,'Muslims'] + df.loc[i,'Christians'] + df.loc[i,'Sikhs'] + df.loc[i,'Buddhists'] + df.loc[i,'Jains'] + df.loc[i,'Others_Religions'] + df.loc[i,'Religion_Not_Stated'])
            elif null_values_filter(df,i,'Workers') == null_values_filter(df,i,'Non_Workers') == False:
                df.loc[i,"Population"] = int(df.loc[i,'Workers'] + df.loc[i,'Non_Workers'])

fill_population(df)

#### Filling up the Male population :

In [ ]:
def fill_male(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Male') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Female') == False:
                df.loc[i,"Male"] = int(df.loc[i,"Population"] - df.loc[i,"Female"])

fill_male(df)

#### Filling up the Female population :

In [ ]:
def fill_female(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Female') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Male') == False:
                df.loc[i,"Female"] = int(df.loc[i,"Population"] - df.loc[i,"Male"])

fill_female(df)

#### Filling up the Literate population :

In [ ]:
def fill_literate(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Literate') == True:
            if null_values_filter(df,i,'Literate_Male') == null_values_filter(df,i,'Literate_Female') == False:
                df.loc[i,"Literate"] = int(df.loc[i,"Literate_Male"] + df.loc[i,"Literate_Female"])

fill_literate(df)

#### Filling up the Literate Male population :

In [ ]:
def fill_literate_male(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Literate_Male') == True:
            if null_values_filter(df,i,'Literate') == null_values_filter(df,i,'Literate_Female') == False:
                df.loc[i,"Literate_Male"] = int(df.loc[i,"Literate"] - df.loc[i,"Literate_Female"])

fill_literate_male(df)

#### Filling up the Literate Female population :

In [ ]:
def fill_literate_female(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Literate_Female') == True:
            if null_values_filter(df,i,'Literate') == null_values_filter(df,i,'Literate_Male') == False:
                df.loc[i,"Literate_Female"] = int(df.loc[i,"Literate"] - df.loc[i,"Literate_Male"])

fill_literate_female(df)

#### Filling up the SC column :

In [ ]:
def fill_sc(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'SC') == True:
            if null_values_filter(df,i,'Male_SC') == null_values_filter(df,i,'Female_SC') == False:
                df.loc[i,"SC"] = int(df.loc[i,"Male_SC"] + df.loc[i,"Female_SC"])

fill_sc(df)

#### Filing up the Male_SC Column :

In [ ]:
def fill_male_sc(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Male_SC') == True:
            if null_values_filter(df,i,'SC') == null_values_filter(df,i,'Female_SC') == False:
                df.loc[i,"Male_SC"] = int(df.loc[i,"SC"] - df.loc[i,"Female_SC"])

fill_male_sc(df)

#### Filing up the Female_SC Column :

In [ ]:
def fill_female_sc(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Female_SC') == True:
            if null_values_filter(df,i,'SC') == null_values_filter(df,i,'Male_SC') == False:
                df.loc[i,'Female_SC'] = int(df.loc[i,'SC'] - df.loc[i,'Male_SC'])

fill_female_sc(df)

#### Filling up the ST column :

In [ ]:
def fill_st(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'ST') == True:
            if null_values_filter(df,i,'Male_ST') == null_values_filter(df,i,'Female_ST') == False:
                df.loc[i,"ST"] = int(df.loc[i,"Male_ST"] + df.loc[i,"Female_ST"])

fill_st(df)

#### Filing up the Male_ST Column :

In [ ]:
def fill_male_st(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Male_ST') == True:
            if null_values_filter(df,i,'ST') == null_values_filter(df,i,'Female_ST') == False:
                df.loc[i,"Male_ST"] = int(df.loc[i,"ST"] - df.loc[i,"Female_ST"])

fill_male_st(df)

#### Filing up the Female_ST Column :

In [ ]:
def fill_female_st(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Female_ST') == True:
            if null_values_filter(df,i,'ST') == null_values_filter(df,i,'Male_ST') == False:
                df.loc[i,'Female_ST'] = int(df.loc[i,'ST'] - df.loc[i,'Male_ST'])

fill_female_st(df)

#### Filling up the Workers column :

In [ ]:
def fill_workers(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Workers') == True:
            if null_values_filter(df,i,'Male_Workers') == null_values_filter(df,i,'Female_Workers') == False:
                df.loc[i,"Workers"] = int(df.loc[i,"Male_Workers"] + df.loc[i,"Female_Workers"])
            elif null_values_filter(df,i,'Main_Workers') == null_values_filter(df,i,'Marginal_Workers') == False:
                df.loc[i,"Workers"] = int(df.loc[i,"Main_Workers"] + df.loc[i,"Marginal_Workers"])
            elif null_values_filter(df,i,'Population') == null_values_filter(df,i,'Non_Workers') == False:
                df.loc[i,"Workers"] = int(df.loc[i,"Population"] - df.loc[i,"Non_Workers"])

fill_workers(df)

#### Filing up the Male_Workers Column :

In [ ]:
def fill_male_workers(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Male_Workers') == True:
            if null_values_filter(df,i,'Workers') == null_values_filter(df,i,'Female_Workers') == False:
                df.loc[i,"Male_Workers"] = int(df.loc[i,"Workers"] - df.loc[i,"Female_Workers"])

fill_male_workers(df)

#### Filing up the Female_Workers Column :

In [ ]:
def fill_female_workers(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Female_Workers') == True:
            if null_values_filter(df,i,'Workers') == null_values_filter(df,i,'Male_Workers') == False:
                df.loc[i,'Female_Workers'] = int(df.loc[i,'Workers'] - df.loc[i,'Male_Workers'])

fill_female_workers(df)

#### Filling up the Main_Workers column :

In [ ]:
def fill_main_workers(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Main_Workers') == True:
            if null_values_filter(df,i,'Workers') == null_values_filter(df,i,'Marginal_Workers') == False:
                df.loc[i,'Main_Workers'] = int(df.loc[i,'Workers'] - df.loc[i,'Marginal_Workers'])
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Non_Workers') == null_values_filter(df,i,'Marginal_Workers') == False:
                df.loc[i,"Main_Workers"] = int(df.loc[i,'Population'] - (df.loc[i,'Non_Workers'] + df.loc[i,'Marginal_Workers']))

fill_main_workers(df)

#### Filling up the Marginal_Workers column :

In [ ]:
def fill_marginal_workers(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Marginal_Workers') == True:
            if null_values_filter(df,i,'Workers') == null_values_filter(df,i,'Main_Workers') == False:
                df.loc[i,'Marginal_Workers'] = int(df.loc[i,'Workers'] - df.loc[i,'Main_Workers'])
            elif null_values_filter(df,i,'Population') == null_values_filter(df,i,'Non_Workers') == null_values_filter(df,i,'Main_Workers') == False:
                df.loc[i,'Marginal_Workers'] = int(df.loc[i,'Population'] - (df.loc[i,'Non_Workers'] + df.loc[i,'Main_Workers']))

fill_marginal_workers(df)

#### Filling up the Non_Workers column:

In [ ]:
def fill_non_workers(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Non_Workers') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Workers') == False:
                df.loc[i,'Non_Workers'] = int(df.loc[i,'Population'] - df.loc[i,'Workers'])
            elif null_values_filter(df,i,'Population') == null_values_filter(df,i,'Main_Workers') == null_values_filter(df,i,'Marginal_Workers') == False:
                df.loc[i,'Non_Workers'] = int(df.loc[i,'Population'] - (df.loc[i,"Main_Workers"] + df.loc[i,"Marginal_Workers"]))

fill_non_workers(df)

#### Filling up the Cultivator_Workers column :

In [ ]:
def fill_cultivator_workers(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Cultivator_Workers') == True:
            if null_values_filter(df,i,'Workers') == null_values_filter(df,i,'Agricultural_Workers') == null_values_filter(df,i,'Household_Workers') == null_values_filter(df,i,'Other_Workers') == False:
                df.loc[i,"Cultivator_Workers"] = int(df.loc[i,'Workers'] - (df.loc[i,'Agricultural_Workers'] + df.loc[i,'Household_Workers'] + df.loc[i,'Other_Workers']))

fill_cultivator_workers(df)

#### Filling up the Agricultural_Workers column :

In [ ]:
def fill_agricultural_workers(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Agricultural_Workers') == True:
            if null_values_filter(df,i,'Workers') == null_values_filter(df,i,'Cultivator_Workers') == null_values_filter(df,i,'Household_Workers') == null_values_filter(df,i,'Other_Workers') == False:
                df.loc[i,"Agricultural_Workers"] = int(df.loc[i,'Workers'] - (df.loc[i,'Cultivator_Workers'] + df.loc[i,'Household_Workers'] + df.loc[i,'Other_Workers']))

fill_agricultural_workers(df)

#### Filling up the Household_Workers column :

In [ ]:
def fill_household_workers(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Household_Workers') == True:
            if null_values_filter(df,i,'Workers') == null_values_filter(df,i,'Cultivator_Workers') == null_values_filter(df,i,'Agricultural_Workers') == null_values_filter(df,i,'Other_Workers') == False:
                df.loc[i,"Household_Workers"] = int(df.loc[i,'Workers'] - (df.loc[i,'Cultivator_Workers'] + df.loc[i,'Agricultural_Workers'] + df.loc[i,'Other_Workers']))

fill_household_workers(df)

### Filling up the Other_Workers column :

In [ ]:
def fill_other_workers(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Other_Workers') == True:
            if null_values_filter(df,i,'Workers') == null_values_filter(df,i,'Cultivator_Workers') == null_values_filter(df,i,'Agricultural_Workers') == null_values_filter(df,i,'Household_Workers') == False:
                df.loc[i,"Other_Workers"] = int(df.loc[i,'Workers'] - (df.loc[i,'Cultivator_Workers'] + df.loc[i,'Agricultural_Workers'] + df.loc[i,'Household_Workers']))

fill_other_workers(df)

#### Filling up the Hindus Column :

In [ ]:
def fill_hindus(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Hindus') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Muslims') == null_values_filter(df,i,'Christians') == null_values_filter(df,i,'Sikhs') == null_values_filter(df,i,'Buddhists') == null_values_filter(df,i,'Jains') == null_values_filter(df,i,'Others_Religions') == null_values_filter(df,i,'Religion_Not_Stated') == False:
                df.loc[i,'Hindus'] = int(df.loc[i,'Population'] - (df.loc[i,'Muslims'] + df.loc[i,'Christians'] + df.loc[i,'Sikhs'] + df.loc[i,'Buddhists'] + df.loc[i,'Jains'] + df.loc[i,'Others_Religions'] + df.loc[i,'Religion_Not_Stated']))

fill_hindus(df)

#### Filling up the Muslims column :

In [ ]:
def fill_muslims(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Muslims') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Hindus') == null_values_filter(df,i,'Christians') == null_values_filter(df,i,'Sikhs') == null_values_filter(df,i,'Buddhists') == null_values_filter(df,i,'Jains') == null_values_filter(df,i,'Others_Religions') == null_values_filter(df,i,'Religion_Not_Stated') == False:
                df.loc[i,'Muslims'] = int(df.loc[i,'Population'] - (df.loc[i,'Hindus'] + df.loc[i,'Christians'] + df.loc[i,'Sikhs'] + df.loc[i,'Buddhists'] + df.loc[i,'Jains'] + df.loc[i,'Others_Religions'] + df.loc[i,'Religion_Not_Stated']))

fill_muslims(df)

### Filling up the Christians column :

In [ ]:
def fill_christians(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Christians') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Hindus') == null_values_filter(df,i,'Muslims') == null_values_filter(df,i,'Sikhs') == null_values_filter(df,i,'Buddhists') == null_values_filter(df,i,'Jains') == null_values_filter(df,i,'Others_Religions') == null_values_filter(df,i,'Religion_Not_Stated') == False:
                df.loc[i,'Christians'] = int(df.loc[i,'Population'] - (df.loc[i,'Hindus'] + df.loc[i,'Muslims'] + df.loc[i,'Sikhs'] + df.loc[i,'Buddhists'] + df.loc[i,'Jains'] + df.loc[i,'Others_Religions'] + df.loc[i,'Religion_Not_Stated']))

fill_christians(df)

#### Filling up the Sikhs column :

In [ ]:
def fill_sikhs(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Sikhs') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Hindus') == null_values_filter(df,i,'Muslims') == null_values_filter(df,i,'Christians') == null_values_filter(df,i,'Buddhists') == null_values_filter(df,i,'Jains') == null_values_filter(df,i,'Others_Religions') == null_values_filter(df,i,'Religion_Not_Stated') == False:
                df.loc[i,'Sikhs'] = int(df.loc[i,'Population'] - (df.loc[i,'Hindus'] + df.loc[i,'Muslims'] + df.loc[i,'Christians'] + df.loc[i,'Buddhists'] + df.loc[i,'Jains'] + df.loc[i,'Others_Religions'] + df.loc[i,'Religion_Not_Stated']))

fill_sikhs(df)

#### Filling up the Buddhists column :

In [ ]:
def fill_buddhists(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Buddhists') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Hindus') == null_values_filter(df,i,'Muslims') == null_values_filter(df,i,'Christians') == null_values_filter(df,i,'Sikhs') == null_values_filter(df,i,'Jains') == null_values_filter(df,i,'Others_Religions') == null_values_filter(df,i,'Religion_Not_Stated') == False:
                df.loc[i,'Buddhists'] = int(df.loc[i,'Population'] - (df.loc[i,'Hindus'] + df.loc[i,'Muslims'] + df.loc[i,'Christians'] + df.loc[i,'Sikhs'] + df.loc[i,'Jains'] + df.loc[i,'Others_Religions'] + df.loc[i,'Religion_Not_Stated']))

fill_buddhists(df)

#### Filling up the Jains column :

In [ ]:
def fill_jains(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Jains') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Hindus') == null_values_filter(df,i,'Muslims') == null_values_filter(df,i,'Christians') == null_values_filter(df,i,'Sikhs') == null_values_filter(df,i,'Buddhists') == null_values_filter(df,i,'Others_Religions') == null_values_filter(df,i,'Religion_Not_Stated') == False:
                df.loc[i,'Jains'] = int(df.loc[i,'Population'] - (df.loc[i,'Hindus'] + df.loc[i,'Muslims'] + df.loc[i,'Christians'] + df.loc[i,'Sikhs'] + df.loc[i,'Buddhists'] + df.loc[i,'Others_Religions'] + df.loc[i,'Religion_Not_Stated']))

fill_jains(df)

#### Filling up the Others_Religions column :

In [ ]:
def fill_others_religions(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Others_Religions') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Hindus') == null_values_filter(df,i,'Muslims') == null_values_filter(df,i,'Christians') == null_values_filter(df,i,'Sikhs') == null_values_filter(df,i,'Buddhists') == null_values_filter(df,i,'Jains') == null_values_filter(df,i,'Religion_Not_Stated') == False:
                df.loc[i,'Others_Religions'] = int(df.loc[i,'Population'] - (df.loc[i,'Hindus'] + df.loc[i,'Muslims'] + df.loc[i,'Christians'] + df.loc[i,'Sikhs'] + df.loc[i,'Buddhists'] + df.loc[i,'Jains'] + df.loc[i,'Religion_Not_Stated']))

fill_others_religions(df)

#### Filling up the Religion_Not_Stated column :

In [ ]:
def fill_religion_not_stated(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Religion_Not_Stated') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Hindus') == null_values_filter(df,i,'Muslims') == null_values_filter(df,i,'Christians') == null_values_filter(df,i,'Sikhs') == null_values_filter(df,i,'Buddhists') == null_values_filter(df,i,'Jains') == null_values_filter(df,i,'Others_Religions') == False:
                df.loc[i,'Religion_Not_Stated'] = int(df.loc[i,'Population'] - (df.loc[i,'Hindus'] + df.loc[i,'Muslims'] + df.loc[i,'Christians'] + df.loc[i,'Sikhs'] + df.loc[i,'Buddhists'] + df.loc[i,'Jains'] + df.loc[i,'Others_Religions']))

fill_religion_not_stated(df)

#### Filling up the Households_Rural column :

In [ ]:
def fill_households_rural(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Households_Rural') == True:
            if null_values_filter(df,i,'Households') == null_values_filter(df,i,'Households_Urban') == False:
                df.loc[i,'Households_Rural'] = int(df.loc[i,"Households"] - df.loc[i,'Households_Urban'])

fill_households_rural(df)

#### Filling up the Households_Urban column :

In [ ]:
def fill_households_urban(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Households_Urban') == True:
            if null_values_filter(df,i,'Households') == null_values_filter(df,i,'Households_Rural') == False:
                df.loc[i,'Households_Urban'] = int(df.loc[i,"Households"] - df.loc[i,'Households_Rural'])

fill_households_urban(df)

#### Filling up the Households column :

In [ ]:
def fill_households(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Households') == True:
            if null_values_filter(df,i,'Households_Urban') == null_values_filter(df,i,'Households_Rural') == False:
                df.loc[i,'Households'] = int(df.loc[i,"Households_Urban"] + df.loc[i,'Households_Rural'])

fill_households(df)


#### Filling up the Literate Education column :

In [ ]:
def fill_literate_education(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Literate_Education') == True:
            if null_values_filter(df,i,'Below_Primary_Education') == null_values_filter(df,i,'Primary_Education') == null_values_filter(df,i,'Middle_Education') == null_values_filter(df,i,'Secondary_Education') == null_values_filter(df,i,'Higher_Education') == null_values_filter(df,i,'Graduate_Education') == null_values_filter(df,i,'Other_Education') == False:
                df.loc[i,'Literate_Education'] = df.loc[i,'Below_Primary_Education'] + df.loc[i,'Primary_Education'] + df.loc[i,'Middle_Education'] + df.loc[i,'Secondary_Education'] + df.loc[i,'Higher_Education'] + df.loc[i,'Graduate_Education'] + df.loc[i,'Other_Education']
            elif null_values_filter(df,i,'Illiterate_Education') == null_values_filter(df,i,'Total_Education') == False:
                df.loc[i,'Literate_Education'] = int(df.loc[i,'Total_Education'] - df.loc[i,'Illiterate_Education'])

fill_literate_education(df)

#### Filling up the Below_Primary_Education column :

In [ ]:
def fill_below_primary_education(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Below_Primary_Education') == True:
            if null_values_filter(df,i,'Literate_Education') == null_values_filter(df,i,'Primary_Education') == null_values_filter(df,i,'Middle_Education') == null_values_filter(df,i,'Secondary_Education') == null_values_filter(df,i,'Higher_Education') == null_values_filter(df,i,'Graduate_Education') == null_values_filter(df,i,'Other_Education') == False:
                df.loc[i,'Below_Primary_Education'] = int(df.loc[i,'Literate_Education'] - (df.loc[i,'Primary_Education'] + df.loc[i,'Middle_Education'] + df.loc[i,'Secondary_Education'] + df.loc[i,'Higher_Education'] + df.loc[i,'Graduate_Education'] + df.loc[i,'Other_Education']))

fill_below_primary_education(df)

#### Filling up the Primary_Education column :

In [ ]:
def fill_primary_education(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Primary_Education') == True:
            if null_values_filter(df,i,'Literate_Education') == null_values_filter(df,i,'Below_Primary_Education') == null_values_filter(df,i,'Middle_Education') == null_values_filter(df,i,'Secondary_Education') == null_values_filter(df,i,'Higher_Education') == null_values_filter(df,i,'Graduate_Education') == null_values_filter(df,i,'Other_Education') == False:
                df.loc[i,'Primary_Education'] = int(df.loc[i,'Literate_Education'] - (df.loc[i,'Below_Primary_Education'] + df.loc[i,'Middle_Education'] + df.loc[i,'Secondary_Education'] + df.loc[i,'Higher_Education'] + df.loc[i,'Graduate_Education'] + df.loc[i,'Other_Education']))

fill_primary_education(df)

#### Filling up the Middle_Education column :

In [ ]:
def fill_middle_education(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Middle_Education') == True:
            if null_values_filter(df,i,'Literate_Education') == null_values_filter(df,i,'Below_Primary_Education') == null_values_filter(df,i,'Primary_Education') == null_values_filter(df,i,'Secondary_Education') == null_values_filter(df,i,'Higher_Education') == null_values_filter(df,i,'Graduate_Education') == null_values_filter(df,i,'Other_Education') == False:
                df.loc[i,'Middle_Education'] = int(df.loc[i,'Literate_Education'] - (df.loc[i,'Below_Primary_Education'] + df.loc[i,'Primary_Education'] + df.loc[i,'Secondary_Education'] + df.loc[i,'Higher_Education'] + df.loc[i,'Graduate_Education'] + df.loc[i,'Other_Education']))

fill_middle_education(df)

#### Filling up the Secondary_Education column :

In [ ]:
def fill_secondary_education(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Secondary_Education') == True:
            if null_values_filter(df,i,'Literate_Education') == null_values_filter(df,i,'Below_Primary_Education') == null_values_filter(df,i,'Primary_Education') == null_values_filter(df,i,'Middle_Education') == null_values_filter(df,i,'Higher_Education') == null_values_filter(df,i,'Graduate_Education') == null_values_filter(df,i,'Other_Education') == False:
                df.loc[i,'Secondary_Education'] = int(df.loc[i,'Literate_Education'] - (df.loc[i,'Below_Primary_Education'] + df.loc[i,'Primary_Education'] + df.loc[i,'Middle_Education'] + df.loc[i,'Higher_Education'] + df.loc[i,'Graduate_Education'] + df.loc[i,'Other_Education']))

fill_secondary_education(df)

#### Filling up the Higher_Education column :

In [ ]:
def fill_higher_education(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Higher_Education') == True:
            if null_values_filter(df,i,'Literate_Education') == null_values_filter(df,i,'Below_Primary_Education') == null_values_filter(df,i,'Primary_Education') == null_values_filter(df,i,'Middle_Education') == null_values_filter(df,i,'Secondary_Education') == null_values_filter(df,i,'Graduate_Education') == null_values_filter(df,i,'Other_Education') == False:
                df.loc[i,'Higher_Education'] = int(df.loc[i,'Literate_Education'] - (df.loc[i,'Below_Primary_Education'] + df.loc[i,'Primary_Education'] + df.loc[i,'Middle_Education'] + df.loc[i,'Secondary_Education'] + df.loc[i,'Graduate_Education'] + df.loc[i,'Other_Education']))

fill_higher_education(df)

#### Filling up the Graduate_Education column :

In [ ]:
def fill_graduate_education(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Graduate_Education') == True:
            if null_values_filter(df,i,'Literate_Education') == null_values_filter(df,i,'Below_Primary_Education') == null_values_filter(df,i,'Primary_Education') == null_values_filter(df,i,'Middle_Education') == null_values_filter(df,i,'Secondary_Education') == null_values_filter(df,i,'Higher_Education') == null_values_filter(df,i,'Other_Education') == False:
                df.loc[i,'Graduate_Education'] = int(df.loc[i,'Literate_Education'] - (df.loc[i,'Below_Primary_Education'] + df.loc[i,'Primary_Education'] + df.loc[i,'Middle_Education'] + df.loc[i,'Secondary_Education'] + df.loc[i,'Higher_Education'] + df.loc[i,'Other_Education']))

fill_graduate_education(df)

#### Filling up the Other_Education column :

In [ ]:
def fill_other_education(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Other_Education') == True:
            if null_values_filter(df,i,'Literate_Education') == null_values_filter(df,i,'Below_Primary_Education') == null_values_filter(df,i,'Primary_Education') == null_values_filter(df,i,'Middle_Education') == null_values_filter(df,i,'Secondary_Education') == null_values_filter(df,i,'Higher_Education') == null_values_filter(df,i,'Graduate_Education') == False:
                df.loc[i,'Other_Education'] = int(df.loc[i,'Literate_Education'] - (df.loc[i,'Below_Primary_Education'] + df.loc[i,'Primary_Education'] + df.loc[i,'Middle_Education'] + df.loc[i,'Secondary_Education'] + df.loc[i,'Higher_Education'] + df.loc[i,'Graduate_Education']))

fill_other_education(df)

#### Filling up the Illiterate_Education column :

In [ ]:
def fill_illiterate_education(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Illiterate_Education') == True:
            if null_values_filter(df,i,'Literate_Education') == null_values_filter(df,i,'Total_Education') == False:
                df.loc[i,'Illiterate_Education'] = int(df.loc[i,'Total_Education'] - df.loc[i,'Literate_Education'])

fill_illiterate_education(df)

#### Filling up the Total_Education column :

In [ ]:
def fill_total_education(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Total_Education') == True:
            if null_values_filter(df,i,'Literate_Education') == null_values_filter(df,i,'Illiterate_Education') == False:
                df.loc[i,'Total_Education'] = int(df.loc[i,'Literate_Education'] + df.loc[i,'Illiterate_Education'])

fill_total_education(df)

#### Filling up the Young_and_Adult column :

In [ ]:
def fill_young_and_adult(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Young_and_Adult') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Middle_Aged') == null_values_filter(df,i,'Senior_Citizen') == null_values_filter(df,i,'Age_Not_Stated') == False:
                df.loc[i,'Young_and_Adult'] = int(df.loc[i,'Population'] - (df.loc[i,'Middle_Aged'] + df.loc[i,'Senior_Citizen'] + df.loc[i,'Age_Not_Stated']))

fill_young_and_adult(df)

#### Filling up the Middle_Aged column :

In [ ]:
def fill_middle_aged(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Middle_Aged') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Young_and_Adult') == null_values_filter(df,i,'Senior_Citizen') == null_values_filter(df,i,'Age_Not_Stated') == False:
                df.loc[i,'Middle_Aged'] = int(df.loc[i,'Population'] - (df.loc[i,'Young_and_Adult'] + df.loc[i,'Senior_Citizen'] + df.loc[i,'Age_Not_Stated']))

fill_middle_aged(df)

#### Filling up the Senior_Citizen column :

In [ ]:
def fill_senior_citizen(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Senior_Citizen') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Young_and_Adult') == null_values_filter(df,i,'Middle_Aged') == null_values_filter(df,i,'Age_Not_Stated') == False:
                df.loc[i,'Senior_Citizen'] = int(df.loc[i,'Population'] - (df.loc[i,'Young_and_Adult'] + df.loc[i,'Middle_Aged'] + df.loc[i,'Age_Not_Stated']))

fill_senior_citizen(df)

#### Filling up the Senior_Citizen column :

In [ ]:
def fill_age_not_stated(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Age_Not_Stated') == True:
            if null_values_filter(df,i,'Population') == null_values_filter(df,i,'Young_and_Adult') == null_values_filter(df,i,'Middle_Aged') == null_values_filter(df,i,'Senior_Citizen') == False:
                df.loc[i,'Age_Not_Stated'] = int(df.loc[i,'Population'] - (df.loc[i,'Young_and_Adult'] + df.loc[i,'Middle_Aged'] + df.loc[i,'Senior_Citizen']))

fill_age_not_stated(df)

#### Filling up the Household_size_1_person_Households column :

In [ ]:
def fill_household_size_1_person_households(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Household_size_1_person_Households') == True:
            if null_values_filter(df,i,'Household_size_1_to_2_persons') == null_values_filter(df,i,'Household_size_2_persons_Households') == False:
                df.loc[i,'Household_size_1_person_Households'] = int(df.loc[i,'Household_size_1_to_2_persons'] - df.loc[i,'Household_size_2_persons_Households'])

fill_household_size_1_person_households(df)

#### Filling up the Household_size_2_persons_Households column :

In [ ]:
def fill_household_size_2_persons_households(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Household_size_2_persons_Households') == True:
            if null_values_filter(df,i,'Household_size_1_to_2_persons') == null_values_filter(df,i,'Household_size_1_person_Households') == False:
                df.loc[i,'Household_size_2_persons_Households'] = int(df.loc[i,'Household_size_1_to_2_persons'] - df.loc[i,'Household_size_1_person_Households'])

fill_household_size_2_persons_households(df)

#### Filling up the Household_size_1_to_2_persons column :

In [ ]:
def fill_household_size_1_to_2_persons(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Household_size_1_to_2_persons') == True:
            if null_values_filter(df,i,'Household_size_1_person_Households') == null_values_filter(df,i,'Household_size_2_persons_Households') == False:
                df.loc[i,'Household_size_1_to_2_persons'] = int(df.loc[i,'Household_size_1_person_Households'] + df.loc[i,'Household_size_2_persons_Households'])

fill_household_size_1_to_2_persons(df)

#### Filling up the Household_size_3_persons_Households column :

In [ ]:
def fill_household_size_3_persons_households(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Household_size_3_persons_Households') == True:
            if null_values_filter(df,i,'Household_size_4_persons_Households') ==  null_values_filter(df,i,'Household_size_5_persons_Households') == null_values_filter(df,i,'Household_size_3_to_5_persons_Households') == False:
                df.loc[i,'Household_size_3_persons_Households'] = int(df.loc[i,'Household_size_3_to_5_persons_Households'] - (df.loc[i,'Household_size_4_persons_Households'] + df.loc[i,'Household_size_5_persons_Households']))

fill_household_size_3_persons_households(df)

#### Filling up the Household_size_4_persons_Households column :

In [ ]:
def fill_household_size_4_persons_households(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Household_size_4_persons_Households') == True:
            if null_values_filter(df,i,'Household_size_3_persons_Households') ==  null_values_filter(df,i,'Household_size_5_persons_Households') == null_values_filter(df,i,'Household_size_3_to_5_persons_Households') == False:
                df.loc[i,'Household_size_4_persons_Households'] = int(df.loc[i,'Household_size_3_to_5_persons_Households'] - (df.loc[i,'Household_size_3_persons_Households'] + df.loc[i,'Household_size_5_persons_Households']))

fill_household_size_4_persons_households(df)

#### Filling up the Household_size_5_persons_Households column :

In [ ]:
def fill_household_size_5_persons_households(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Household_size_5_persons_Households') == True:
            if null_values_filter(df,i,'Household_size_3_persons_Households') ==  null_values_filter(df,i,'Household_size_4_persons_Households') == null_values_filter(df,i,'Household_size_3_to_5_persons_Households') == False:
                df.loc[i,'Household_size_5_persons_Households'] = int(df.loc[i,'Household_size_3_to_5_persons_Households'] - (df.loc[i,'Household_size_3_persons_Households'] + df.loc[i,'Household_size_4_persons_Households']))

fill_household_size_5_persons_households(df)

#### Filling up the Household_size_3_to_5_persons_Households column :

In [ ]:
def fill_household_size_3_to_5_persons_households(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Household_size_3_to_5_persons_Households') == True:
            if null_values_filter(df,i,'Household_size_3_persons_Households') == null_values_filter(df,i,'Household_size_4_persons_Households') == null_values_filter(df,i,'Household_size_5_persons_Households') == False:
                df.loc[i,'Household_size_3_to_5_persons_Households'] = int(df.loc[i,'Household_size_3_persons_Households'] + df.loc[i,'Household_size_4_persons_Households'] + df.loc[i,'Household_size_5_persons_Households'])

fill_household_size_3_to_5_persons_households(df)

#### Filling up the Power_Parity_Rs_45000_90000 column :

In [ ]:


def fill_power_parity_rs_45000_90000(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == True:
            if null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == null_values_filter(df,i,'Power_Parity_Rs_45000_150000') == False:
                df.loc[i,'Power_Parity_Rs_45000_90000'] = int(df.loc[i,'Power_Parity_Rs_45000_150000'] - df.loc[i,'Power_Parity_Rs_90000_150000'])
            elif null_values_filter(df,i,'Total_Power_Parity') == null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == False:
                df.loc[i,'Power_Parity_Rs_45000_90000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Less_than_Rs_45000'] + df.loc[i,'Power_Parity_Rs_90000_150000'] + df.loc[i,'Power_Parity_Rs_150000_240000'] + df.loc[i,'Power_Parity_Rs_240000_330000'] + df.loc[i,'Power_Parity_Rs_330000_425000'] + df.loc[i,'Power_Parity_Rs_425000_545000'] + df.loc[i,'Power_Parity_Above_Rs_545000']))

fill_power_parity_rs_45000_90000(df)

#### Filling up the Power_Parity_Rs_90000_150000 column :

In [ ]:
def fill_power_parity_rs_90000_150000(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == True:
            if null_values_filter(df,i,'Power_Parity_Rs_45000_150000') == null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == False:
                df.loc[i,'Power_Parity_Rs_90000_150000'] = int(df.loc[i,'Power_Parity_Rs_45000_150000'] - df.loc[i,'Power_Parity_Rs_45000_90000'])
            elif null_values_filter(df,i,'Total_Power_Parity') == null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == False:
                df.loc[i,'Power_Parity_Rs_90000_150000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Less_than_Rs_45000'] + df.loc[i,'Power_Parity_Rs_45000_90000'] + df.loc[i,'Power_Parity_Rs_150000_240000'] + df.loc[i,'Power_Parity_Rs_240000_330000'] + df.loc[i,'Power_Parity_Rs_330000_425000'] + df.loc[i,'Power_Parity_Rs_425000_545000'] + df.loc[i,'Power_Parity_Above_Rs_545000']))

fill_power_parity_rs_90000_150000(df)

#### Filling up the Power_Parity_Rs_45000_150000 column :

In [ ]:
def fill_power_parity_rs_45000_150000(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Power_Parity_Rs_45000_150000') == True:
            if null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == False:
                df.loc[i,'Power_Parity_Rs_45000_150000'] = int(df.loc[i,'Power_Parity_Rs_45000_90000'] + df.loc[i,'Power_Parity_Rs_90000_150000'])
            elif null_values_filter(df,i,'Total_Power_Parity') == null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == False:
                df.loc[i,'Power_Parity_Rs_45000_150000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Less_than_Rs_45000'] + df.loc[i,'Power_Parity_Rs_150000_240000'] + df.loc[i,'Power_Parity_Rs_240000_330000'] + df.loc[i,'Power_Parity_Rs_330000_425000'] + df.loc[i,'Power_Parity_Rs_425000_545000'] + df.loc[i,'Power_Parity_Above_Rs_545000']))

fill_power_parity_rs_45000_150000(df)

#### Filling up the Power_Parity_Rs_150000_240000 column :

In [ ]:
def fill_power_parity_rs_150000_240000(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == True:
            if null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == null_values_filter(df,i,'Power_Parity_Rs_150000_330000') == False:
                df.loc[i,'Power_Parity_Rs_150000_240000'] = int(df.loc[i,'Power_Parity_Rs_150000_330000'] - df.loc[i,'Power_Parity_Rs_240000_330000'])
            elif null_values_filter(df,i,'Total_Power_Parity') == null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == False:
                df.loc[i,'Power_Parity_Rs_150000_240000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Less_than_Rs_45000'] + df.loc[i,'Power_Parity_Rs_45000_90000'] + df.loc[i,'Power_Parity_Rs_90000_150000'] + df.loc[i,'Power_Parity_Rs_240000_330000'] + df.loc[i,'Power_Parity_Rs_330000_425000'] + df.loc[i,'Power_Parity_Rs_425000_545000'] + df.loc[i,'Power_Parity_Above_Rs_545000']))

fill_power_parity_rs_150000_240000(df)

#### Filling up the Power_Parity_Rs_240000_330000 column :

In [ ]:
def fill_power_parity_rs_240000_330000(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == True:
            if null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == null_values_filter(df,i,'Power_Parity_Rs_150000_330000') == False:
                df.loc[i,'Power_Parity_Rs_240000_330000'] = int(df.loc[i,'Power_Parity_Rs_150000_330000'] - df.loc[i,'Power_Parity_Rs_150000_240000'])
            elif null_values_filter(df,i,'Total_Power_Parity') == null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == False:
                df.loc[i,'Power_Parity_Rs_240000_330000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Less_than_Rs_45000'] + df.loc[i,'Power_Parity_Rs_45000_90000'] + df.loc[i,'Power_Parity_Rs_90000_150000'] + df.loc[i,'Power_Parity_Rs_150000_240000'] + df.loc[i,'Power_Parity_Rs_330000_425000'] + df.loc[i,'Power_Parity_Rs_425000_545000'] + df.loc[i,'Power_Parity_Above_Rs_545000']))

fill_power_parity_rs_240000_330000(df)

#### Filling up the Power_Parity_Rs_150000_330000 column :

In [ ]:
def fill_power_parity_rs_150000_330000(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Power_Parity_Rs_150000_330000') == True:
            if null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == False:
                df.loc[i,'Power_Parity_Rs_150000_330000'] = int(df.loc[i,'Power_Parity_Rs_150000_240000'] + df.loc[i,'Power_Parity_Rs_240000_330000'])
            elif null_values_filter(df,i,'Total_Power_Parity') == null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == False:
                df.loc[i,'Power_Parity_Rs_150000_330000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Less_than_Rs_45000'] + df.loc[i,'Power_Parity_Rs_45000_90000'] + df.loc[i,'Power_Parity_Rs_90000_150000'] + df.loc[i,'Power_Parity_Rs_330000_425000'] + df.loc[i,'Power_Parity_Rs_425000_545000'] + df.loc[i,'Power_Parity_Above_Rs_545000']))

fill_power_parity_rs_150000_330000(df)

#### Filling up the Power_Parity_Rs_330000_425000 column :

In [ ]:
def fill_power_parity_rs_330000_425000(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == True:
            if null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == null_values_filter(df,i,'Power_Parity_Rs_330000_545000') == False:
                df.loc[i,'Power_Parity_Rs_330000_425000'] = int(df.loc[i,'Power_Parity_Rs_330000_545000'] - df.loc[i,'Power_Parity_Rs_425000_545000'])
            elif null_values_filter(df,i,'Total_Power_Parity') == null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == False:
                df.loc[i,'Power_Parity_Rs_330000_425000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Less_than_Rs_45000'] + df.loc[i,'Power_Parity_Rs_45000_90000'] + df.loc[i,'Power_Parity_Rs_90000_150000'] + df.loc[i,'Power_Parity_Rs_150000_240000'] + df.loc[i,'Power_Parity_Rs_240000_330000'] + df.loc[i,'Power_Parity_Rs_425000_545000'] + df.loc[i,'Power_Parity_Above_Rs_545000']))

fill_power_parity_rs_330000_425000(df)

#### Filling up the Power_Parity_Rs_425000_545000 column :

In [ ]:
def fill_power_parity_rs_425000_545000(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == True:
            if null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == null_values_filter(df,i,'Power_Parity_Rs_330000_545000') == False:
                df.loc[i,'Power_Parity_Rs_425000_545000'] = int(df.loc[i,'Power_Parity_Rs_330000_545000'] - df.loc[i,'Power_Parity_Rs_330000_425000'])
            elif null_values_filter(df,i,'Total_Power_Parity') == null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == False:
                df.loc[i,'Power_Parity_Rs_425000_545000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Less_than_Rs_45000'] + df.loc[i,'Power_Parity_Rs_45000_90000'] + df.loc[i,'Power_Parity_Rs_90000_150000'] + df.loc[i,'Power_Parity_Rs_150000_240000'] + df.loc[i,'Power_Parity_Rs_240000_330000'] + df.loc[i,'Power_Parity_Rs_330000_425000'] + df.loc[i,'Power_Parity_Above_Rs_545000']))

fill_power_parity_rs_425000_545000(df)

#### Filling up the Power_Parity_Rs_330000_545000 :

In [ ]:
def fill_power_parity_rs_330000_545000(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Power_Parity_Rs_330000_545000') == True:
            if null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == False:
                df.loc[i,'Power_Parity_Rs_330000_545000'] = int(df.loc[i,'Power_Parity_Rs_330000_425000'] + df.loc[i,'Power_Parity_Rs_425000_545000'])
        elif null_values_filter(df,i,'Total_Power_Parity') == null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == False:
                df.loc[i,'Power_Parity_Rs_330000_545000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Less_than_Rs_45000'] + df.loc[i,'Power_Parity_Rs_45000_90000'] + df.loc[i,'Power_Parity_Rs_90000_150000'] + df.loc[i,'Power_Parity_Rs_150000_240000'] + df.loc[i,'Power_Parity_Rs_240000_330000'] + df.loc[i,'Power_Parity_Above_Rs_545000']))

fill_power_parity_rs_330000_545000(df)

#### Filling up the Total_Power_Parity column :

In [ ]:
def fill_total_power_parity(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Total_Power_Parity') == True:
            if null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_45000_150000') == null_values_filter(df,i,'Power_Parity_Rs_150000_330000') == null_values_filter(df,i,'Power_Parity_Rs_330000_545000') == null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == False:
                df.loc[i,'Total_Power_Parity'] = int(df.loc[i,'Power_Parity_Less_than_Rs_45000'] + df.loc[i,'Power_Parity_Rs_45000_150000'] + df.loc[i,'Power_Parity_Rs_150000_330000'] + df.loc[i, 'Power_Parity_Rs_330000_545000'] + df.loc[i,'Power_Parity_Above_Rs_545000'])
            elif null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == False:
                df.loc[i,'Total_Power_Parity'] = int(df.loc[i,'Power_Parity_Less_than_Rs_45000'] + df.loc[i,'Power_Parity_Rs_45000_90000'] + df.loc[i,'Power_Parity_Rs_90000_150000'] + df.loc[i,'Power_Parity_Rs_150000_240000'] + df.loc[i,'Power_Parity_Rs_240000_330000'] + df.loc[i,'Power_Parity_Rs_330000_425000'] + df.loc[i,'Power_Parity_Rs_425000_545000'] + df.loc[i,'Power_Parity_Above_Rs_545000'])

fill_total_power_parity(df)

#### Filling up the Power_Parity_Less_than_Rs_45000 column :

In [ ]:

def fill_power_parity_less_than_rs_45000(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == True:
            if null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == null_values_filter(df,i,'Power_Parity_Rs_45000_150000') == null_values_filter(df,i,'Power_Parity_Rs_150000_330000') == null_values_filter(df,i,'Power_Parity_Rs_330000_545000') == null_values_filter(df,i,'Total_Power_Parity') == False:
                df.loc[i,'Power_Parity_Less_than_Rs_45000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Above_Rs_545000'] + df.loc[i,'Power_Parity_Rs_45000_150000'] + df.loc[i,'Power_Parity_Rs_150000_330000'] + df.loc[i, 'Power_Parity_Rs_330000_545000']))
            elif null_values_filter(df,i,'Total_Power_Parity') == null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == False:
                df.loc[i,'Power_Parity_Less_than_Rs_45000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Rs_45000_90000'] + df.loc[i,'Power_Parity_Rs_90000_150000'] + df.loc[i,'Power_Parity_Rs_150000_240000'] + df.loc[i,'Power_Parity_Rs_240000_330000'] + df.loc[i,'Power_Parity_Rs_330000_425000'] + df.loc[i,'Power_Parity_Rs_425000_545000'] + df.loc[i,'Power_Parity_Above_Rs_545000']))

fill_power_parity_less_than_rs_45000(df)

#### Filling up the Power_Parity_Above_Rs_545000 column :

In [ ]:
def fill_power_parity_above_rs_545000(df):
    for i in range(len(df)):
        if null_values_filter(df,i,'Power_Parity_Above_Rs_545000') == True:
            if null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_45000_150000') == null_values_filter(df,i,'Power_Parity_Rs_150000_330000') == null_values_filter(df,i,'Power_Parity_Rs_330000_545000') == null_values_filter(df,i,'Total_Power_Parity') == False:
                df.loc[i,'Power_Parity_Above_Rs_545000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Less_than_Rs_45000'] + df.loc[i,'Power_Parity_Rs_45000_150000'] + df.loc[i,'Power_Parity_Rs_150000_330000'] + df.loc[i, 'Power_Parity_Rs_330000_545000']))
            elif null_values_filter(df,i,'Total_Power_Parity') == null_values_filter(df,i,'Power_Parity_Less_than_Rs_45000') == null_values_filter(df,i,'Power_Parity_Rs_45000_90000') == null_values_filter(df,i,'Power_Parity_Rs_90000_150000') == null_values_filter(df,i,'Power_Parity_Rs_150000_240000') == null_values_filter(df,i,'Power_Parity_Rs_240000_330000') == null_values_filter(df,i,'Power_Parity_Rs_330000_425000') == null_values_filter(df,i,'Power_Parity_Rs_425000_545000') == False:
                df.loc[i,'Power_Parity_Above_Rs_545000'] = int(df.loc[i,'Total_Power_Parity'] - (df.loc[i,'Power_Parity_Rs_45000_90000'] + df.loc[i,'Power_Parity_Rs_90000_150000'] + df.loc[i,'Power_Parity_Rs_150000_240000'] + df.loc[i,'Power_Parity_Rs_240000_330000'] + df.loc[i,'Power_Parity_Rs_330000_425000'] + df.loc[i,'Power_Parity_Rs_425000_545000'] + df.loc[i,'Power_Parity_Less_than_Rs_45000']))

fill_power_parity_above_rs_545000(df)

###  Finding the percentage of missing data after cleaning is done ...

In [ ]:
missing_data_after_cleaning = missingDataPercent(df)

## Creating a CSV file to store the missing data details to visualize it in the streamlit along with other queries :

In [ ]:
missing_data = {
    'Columns': df.columns,
    'Missing_data_before_cleaning (%)' : missing_data_before_cleaning.values(),
    'Missing_data_after_cleaning (%)' : missing_data_after_cleaning.values()
    }
missing_data_df = pd.DataFrame(missing_data)

missing_data_df.to_csv('missing_data.csv', header=True)

### Visualizing missing data Before & After Cleaning :

In [92]:
missing_data_df

Columns  Missing_data_before_cleaning (%)  \
0                    District code                              0.00   
1                         State/UT                              0.00   
2                         District                              0.00   
3                       Population                              4.69   
4                             Male                              4.69   
..                             ...                               ...   
113  Power_Parity_Rs_330000_425000                              5.16   
114  Power_Parity_Rs_425000_545000                              4.69   
115  Power_Parity_Rs_330000_545000                              3.59   
116   Power_Parity_Above_Rs_545000                              4.69   
117             Total_Power_Parity                              5.00   

     Missing_data_after_cleaning (%)  
0                               0.00  
1                               0.00  
2                               0.00  
3                               0.00  
4                               0.00  
..                               ...  
113                             0.16  
114                             0.31  
115                             0.16  
116                             0.62  
117                             0.94  

[118 rows x 3 columns]

In [96]:
from google.colab import sheets
sheet = sheets.InteractiveSheet(df=missing_data_df)

ModuleNotFoundError: No module named 'google.colab'

### Task -5 : Save Data to MongoDB

In [ ]:
# Connecting with the MongoDB Locallhost 

uri = "mongodb+srv://sashikantaraja1966:12345@cluster0.ndy0rxz.mongodb.net/"

def create_mongo_connection(uri):

    # Create a new client and connect to the server
    client = MongoClient(uri, server_api=ServerApi('1'))

    # Send a ping to confirm a successful connection
    try:
        client.admin.command('ping')
        print("Pinged your deployment. You successfully connected to MongoDB!")
    except Exception as e:
        print(e)

create_mongo_connection(uri)

Pinged your deployment. You successfully connected to MongoDB!


### Creating a DB and Collection.

In [102]:
# Creating a DB and Collection

def mongo_create_db_collection(uri):
    client=MongoClient(f'{uri}').census_db
    collection = 'census'

    # Checking if the collection already exists
    if collection in client.list_collection_names():
        client[collection].drop()
        client=MongoClient(f'{uri}').census_db.census
    else:
        client=MongoClient(f"{uri}").census_db.census

    return client

client = mongo_create_db_collection(uri)

### Inserting the data into MongoDB *>

In [ ]:
def mongo_insert(client,df):
    keys = list(df.columns)
    for i in range(len(df)):
        values = []
        for j in keys:
            if j not in ['State/UT','District']:
                try:
                    values.append(int(df.loc[i,j]))
                except:
                    values.append(None)
            else:
                values.append(df.loc[i,j])
        doc = dict(zip(keys,values))
        client.insert_one(doc)

mongo_insert(client,df)

### Task 6: Database connection and data upload

In [97]:
# Connecting to MySQL server Localhost

mydb = mysql.connector.connect(
 host="localhost",
 user="root",
 password="",)

print(mydb)
mycursor = mydb.cursor(buffered=True)

### Creating DB and & Manually Giving Names :

- Manually giving the names for columns which are more than 65 characters in length as MySQL column names have certain constraints
over Column name's length and special character usage.

In [ ]:

def create_db_table(df):
    mycursor.execute('CREATE DATABASE IF NOT EXISTS Census_DB')

    columns = list()
    for i in df.columns:
        if i == 'District code':
            columns.append('District_Code INT PRIMARY KEY')
        elif i == "State/UT":
            columns.append('State_UT VARCHAR(50)')
        elif i == "District":
            columns.append('District VARCHAR(50)')
        elif i == "Households_with_TV_Computer_Laptop_Telephone_mobile_phone_and_Scooter_Car":
            columns.append("Households_TV_Computer_Laptop_Telephone_mobile_phone_Scooter_Car INT")
        elif i == "Type_of_latrine_facility_Night_soil_disposed_into_open_drain_Households":
            columns.append("Type_of_latrine_facility_Night_soil_disposed_into_open_drain INT")
        elif i == "Type_of_latrine_facility_Flush_pour_flush_latrine_connected_to_other_system_Households":
            columns.append("Type_of_latrine_Flush_pour_connected_to_other_system_Households INT")
        elif i == "Not_having_latrine_facility_within_the_premises_Alternative_source_Open_Households":
            columns.append("Not_having_latrine_within_premises_Other_source_Open_Households INT")
        elif i == "Main_source_of_drinking_water_Handpump_Tubewell_Borewell_Households":
            columns.append("Source_of_drinking_water_Handpump_Tubewell_Borewell_Households INT")
        elif i == "Main_source_of_drinking_water_Other_sources_Spring_River_Canal_Tank_Pond_Lake_Other_sources__Households":
            columns.append("Drinking_water_Spring_River_Canal_Tank_Pond_Lake_Other_Household INT")
        else:
            columns.append(f'{i} INT')

    mycursor.execute(f'CREATE OR REPLACE TABLE Census_DB.census_2011 ({", ".join(columns)})')
    mydb.commit()

create_db_table(df)

### Getting the column names from the table:

In [ ]:
def get_columns(table_name):
    mycursor.execute(f"DESCRIBE {table_name}")
    table_columns = []
    for i in mycursor:
        table_columns.append(i[0])

    return tuple(table_columns)

table_columns = get_columns('Census_DB.census_2011')

### Creating Function to insert data into DB

- Creating placeholders for the values as we have to insert "Null" values inside the table and SQL accepts
string formatting in '%s' method.

- Using execute many to insert bulk data into table.

In [104]:
def db_insert_data(mongo_client, df, table_columns):
    placeholders = ", ".join(["%s"] * len(table_columns))
    columns = ", ".join(table_columns)

    data = []

    for i in range(1,len(df)+1):
        for j in mongo_client.find({'District code':i}):
            data.append(tuple(j.values())[1:])

    try :
        query = f'INSERT INTO Census_DB.Census_2011 ({columns}) VALUES ({placeholders})'
        mycursor.executemany(query, data)
    except Exception as e:
        print(e)

    mydb.commit()

db_insert_data(client, df, table_columns)